### PR TITLE
Register a Config instance to bootstrap tests

### LINKED ISSUE #5702: Reexamine how we manage config in tests
**Description**
We should pin down our configuration lifecycle.  In getting #5387 to pass, I ran across several cases where the configuration is referenced before it is actually set up by our loader.  This included `TestResourceManager`. I've added in code to explicitly set up a temporary configuration, but we should look at other options, such as setting up the build time config object much earlier in the process.

It needs to be determined what the expectation is for tests using configuration.  Class setup actions for example are unlikely to be able to access run time config meaningfully.

See https://github.com/quarkusio/quarkus/pull/5387#discussion_r349641331 for discussion.

### LINKED ISSUE #35693: Confusing documentation or behaviour for `quarkus.test.profile.tags`
### Describe the bug

The [configurations options page](https://quarkus.io/guides/all-config#quarkus-core_quarkus.test.profile.tags) mentions that `quarkus.test.profile.tags` controls which tests will be run or not. It also mentions an environment variable name that can be used to set this value. This seems to suggest that this value can be controlled in any of the "standard" ways configuration is handled in Quarkus: through yaml files, environment variables, etc.

However, it turns out that this value is _only_ read from the system properties directly ([source](https://github.com/quarkusio/quarkus/blob/00ebf1241cecf251790c3e1142cf362cf9b5e014/test-framework/junit5/src/main/java/io/quarkus/test/junit/QuarkusTestExtension.java#L1193)), and thus setting this value through an environment variable does not work. Additionally, for some reason, using `-Dquarkus.test.profile.tags=xyz` in my `gradle test` command also does not seem to pass this property on after the fork. So for me, the only reason to set this field now it by hardcoding it in the Gradle task configuration as outlined [here](https://quarkus.io/guides/gradle-tooling#custom-test-configuration-profile). This seems very limited and is quite inconvenient.

### Expected behavior

Ideally, based on the documentation, I would expect this configuration to work through all normal Quarkus config options. However, if that is not possible (due to startup dependencies - I can imagine), then at the very least I would expect the documentation to be more explicit about this. Ideally, then, the respective gradle property should be passed on to the test runtime JVM as well, to ease the use of this setting.

### Actual behavior

_No response_

### How to Reproduce?

_No response_

### Output of `uname -a` or `ver`

_No response_

### Output of `java -version`

openjdk version "17.0.8.1"

### GraalVM version (if different from Java)

_No response_

### Quarkus version or git rev

3.3.0

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

Gradle 8.1.1

### Additional information

_No response_

### LINKED ISSUE #45316: quarkus.profile is 'prod' when running tests not marked with @QuarkusTest while having junit-platform.properties file
### Describe the bug

Creating the file `src/test/resources/junit-platform.properties` will make `quarkus.profile=prod` as default when running tests. This is not the case for tests that are annotated with @QuarkusTest.

### Expected behavior

Running tests should have `quarkus.profile=test` as default

### Actual behavior

Tests (with the exception of those marked with @QuarkusTest) have `quarkus.profile=prod` as default

### How to Reproduce?

1. Generate a new barebones quarkus application from https://code.quarkus.io/
2. Create a test class like: 
```java 
import org.eclipse.microprofile.config.ConfigProvider;
import org.junit.jupiter.api.Test;

import static org.junit.jupiter.api.Assertions.assertEquals;

public class ProfileTest {
    @Test
    void profileShouldBeTest() {
        final String profile = ConfigProvider.getConfig().getValue("quarkus.profile", String.class);
        assertEquals("test", profile);
    }
}
```
3. Verify the test succeeds
4. Add an empty file `src/test/resources/junit-platform.properties`
5. Run test again and notice that it fails: `Actual   :prod`
6. Add `@io.quarkus.test.junit.QuarkusTest` annotation to the class and see that it succeeds again

Note: annotating the class with @io.quarkus.test.junit.QuarkusIntegrationTest does not make the test succeed.

### Output of `uname -a` or `ver`

Microsoft Windows [Version 10.0.22631.4602]

### Output of `java -version`

openjdk version "21.0.2" 2024-01-16 OpenJDK Runtime Environment (build 21.0.2+13-58) OpenJDK 64-Bit Server VM (build 21.0.2+13-58, mixed mode, sharing)

### Quarkus version or git rev

3.17.5

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

Apache Maven 3.9.9 (8e8579a9e76f7d015ee5ec7bfcdc97d260186937) Maven home: C:\Program Files\.....\maven\lib\maven3 Java version: 21.0.2, vendor: Oracle Corporation, runtime: C:\Users\.....\.jdks\openjdk-21.0.2 Default locale: en_GB, platform encoding: UTF-8 OS name: "windows 11", version: "10.0", arch: "amd64", family: "windows"

### Additional information

_No response_

### PR DESCRIPTION
**Pull request by radcortez**
The goal of this PR is to bootstrap a `Config` instance very early in the testing process so it can be used in testing setup code, like in JUnit Extensions and listeners.

There is a new module called `quarkus-junit5-config`, replacing `quarkus-junit5-properties`. Right now, it is limited, but it can be expanded to support many of the JUnit configurations using our mechanisms.

We register the `Config` in the test classloader with a JUnit `LauncherSessionListener`. A call to `ConfigProvider.getConfig` can either return the test config or the Quarkus config, depending on the classloader. It is now consistent, whereas before, it could depend on where it was called (because `ConfigProvider.getConfig` would also create the config if unavailable). 

Our config is now more tightly integrated with JUnit, so we don't require JUnit properties file to configure the `QuarkusTestProfileAwareClassOrderer`. We add a global JUnit `ClassOrderer` that can delegate to any other `ClassOrderer` provided by the Quarkus configuration. 

We also redid the logging configuration on test. Since we now have a reliable configuration, there is no need to juggle the config as we did in `BasicLoggingEnabler`, and there is a simple `LoggingSetupExtension` that sets up the log.

Since config is more consistent with the testing code, we could remove many of the workarounds we had. For example:
- https://github.com/quarkusio/quarkus/pull/42715/files#diff-6cb808317e81756dc4ed50ac2bfade53d657ca3bdc3ce753a1e5c24b5f2ecf68L150-L152
- https://github.com/quarkusio/quarkus/pull/42715/files#diff-ec23461096492124cf4ddf4b01c898881c2c57d3150e327f1e00b3e7e81564e6L2-L3
- https://github.com/quarkusio/quarkus/pull/42715/files#diff-0976161629e37650ede7638425bb54845838f7aeb4fc3fae3e30a66153411784L41-L43
- https://github.com/quarkusio/quarkus/pull/42715/files#diff-a050ea6d2087e6491a6cfcadd0ef58e0ffbe09935ab12850ca3adbc38ce14bffL217-L228

There are still more workarounds spread through the code that we can remove after.

- Fixes #45316
- Fixes #35693
- Fixes #5702

Breaking Changes:

- Because Quarkus sets its own `ClassOrderer`, if a user wanted to supply its own, it would require to exclude `io.quarkus:quarkus-junit5-properties` from the test dependencies and provide a `junit-platform.properties` with the property `junit.jupiter.testclass.order.default` and FQN of the ordered as value. With this PR, the user only has to provide the orderer's FQN in `quarkus.test.class-orderer`.

### CODE REVIEW

**Comment by geoand** [comment]
This is really cool!

I assume the CI failures are related :)

**Comment by radcortez** [comment]
Strange. These were ok before. It was probably a last minute change. I'll have a look.

**Comment by geoand** [comment]
@holly-cummins will likely be interested in this as well

File: core/deployment/src/main/java/io/quarkus/deployment/dev/testing/TestConfig.java [hunk 1]
```diff
@@ -45,6 +47,15 @@ public interface TestConfig {
...
     boolean displayTestOutput();
 
+    /**
+     * The FQN of the JUnit ClassOrdered to use. If the class cannot be found, it defaults backs to JUnit class order.
+     *
+     * @see <a href=https://junit.org/junit5/docs/current/user-guide/#writing-tests-test-execution-order-classes>JUnit Class
+     *      Order<a/>
+     */
+    @WithDefault("io.quarkus.test.junit.util.QuarkusTestProfileAwareClassOrderer")
+    Optional<String> classOrderer();
```
Comment (geoand) [inline core/deployment/src/main/java/io/quarkus/deployment/dev/testing/TestConfig.java:57]: Would our JUnit extension continue to work if this is changed?

**Review by geoand — COMMENTED** [review]
Very nice work @radcortez!

Apart from the one question I asked in the review, I would also like to know if you have seen in class loading related leaks with this change.
Asking because the test support is often a culprit for such issues

File: core/deployment/src/main/java/io/quarkus/deployment/dev/testing/TestConfig.java (hunk 1 above)
Comment (radcortez) [reply core/deployment/src/main/java/io/quarkus/deployment/dev/testing/TestConfig.java:57]: Yes. The `QuarkusTestProfileAwareClassOrderer` is used to optimize the test run order and minimize Quarkus restarts when executing tests: https://github.com/quarkusio/quarkus/pull/20156

It is also documented here:
https://quarkus.io/version/3.15/guides/getting-started-testing#testing_different_profiles

Users can replace ours and provide their own, but setting that up is a bit cumbersome (JUnit only supports a single config file, which we provide and you have to exclude dependencies).

What I did is to provide a global empty one that can use our configuration to delegate to it:
https://github.com/quarkusio/quarkus/pull/42715/files#diff-8f3e9670c7afd7364f77c048342c4973cab00e10963299434315f736f348ddc8R13-R54

By default, you get `QuarkusTestProfileAwareClassOrderer`. To get rid of it or override it, we only need to set `quarks.test.class-orderer`. This is actually a breaking change.

Comment (geoand) [reply core/deployment/src/main/java/io/quarkus/deployment/dev/testing/TestConfig.java:57]: Right. What I am worried about is that if users use their own, nothing will work - which is unexpected and we should definitely tell users what to do.

**Comment by radcortez** [comment]
> Apart from the one question I asked in the review, I would also like to know if you have seen in class loading related leaks with this change.

I didn't. It doesn't mean that it is not there :)

I've tried to avoid messing with the CL as much as possible. A `Config` registers per CL, but that was already the case. We just didn't know when that would happen because it would be on the first call of `ConfigProvider.getConfig`. With this change, it would always happen in the same place, making it consistent.

There is one change that was a bit tricky, which is:
https://github.com/quarkusio/quarkus/pull/42715/files#diff-462fa0cb443c21c9406c8b4daeb31500f20a4c9b46a0b039adcc93dee6a3bea4R265-R267

The `ConfigProviderResolver` contains a `Map` with a `Config` per CL. Actually, I think I'm missing a release there. Let me add it :)

File: core/deployment/src/main/java/io/quarkus/deployment/dev/testing/TestConfig.java (hunk 1 above)
Comment (radcortez) [reply core/deployment/src/main/java/io/quarkus/deployment/dev/testing/TestConfig.java:57]: Yes, we need to add a note to use `quarks.test.class-orderer` instead.

**Comment by geoand** [comment]
I would appreciate it a lot if you could run one of the tests that contains multiple test profiles (which force a restart) with and without this PR to make sure that it has not introduced some subtle leak.

File: core/deployment/src/main/java/io/quarkus/deployment/dev/testing/TestConfig.java (hunk 1 above)
Comment (geoand) [reply core/deployment/src/main/java/io/quarkus/deployment/dev/testing/TestConfig.java:57]: 👍🏽

**Comment by radcortez** [comment]
> The `LauncherInterceptor` is invoked a few times during test launch, but the one I've been looking at is before discovery. The class orderer is invoked after discovery. I'm not sure if that would affect the success of what you're doing.

I had a better look, and I'm now using a `LauncherSessionListener` to implement the logic, which seems to execute after the first interceptor call. It works just as well. This should allow you to change any Classloader stuff before we set the configuration (where we need the Classloader we want to use).

**Review by geoand — APPROVED** [review]
Awesome work @radcortez !

**Review by gsmet — COMMENTED** [review]
This looks like very nice cleanup, thanks!

I added a few questions and some micro suggestions for the doc.

File: test-framework/junit5/src/main/java/io/quarkus/test/junit/util/QuarkusTestProfileAwareClassOrderer.java [hunk 2]
```diff
@@ -69,31 +71,57 @@ public class QuarkusTestProfileAwareClassOrderer implements ClassOrderer {
...
-                CFGKEY_ORDER_PREFIX_QUARKUS_TEST_WITH_RESTRICTED_RES,
-                DEFAULT_ORDER_PREFIX_QUARKUS_TEST_WITH_RESTRICTED_RES,
-                context);
-        var prefixNonQuarkusTest = getConfigParam(
-                CFGKEY_ORDER_PREFIX_NON_QUARKUS_TEST,
-                DEFAULT_ORDER_PREFIX_NON_QUARKUS_TEST,
-                context);
 
         // first pass: run secondary orderer first (!), which is easier than running it per "grouping"
-        buildSecondaryOrderer(context).orderClasses(context);
```
Comment (famod) [inline test-framework/junit5/src/main/java/io/quarkus/test/junit/util/QuarkusTestProfileAwareClassOrderer.java:96]: AFAICS `buildSecondaryOrderer()` is not called anymore so it should be removed.

File: core/deployment/src/main/java/io/quarkus/deployment/dev/testing/TestConfig.java [hunk 3]
```diff
@@ -45,6 +47,15 @@ public interface TestConfig {
     @WithDefault("false")
     boolean displayTestOutput();
 
+    /**
+     * The FQN of the JUnit ClassOrdered to use. If the class cannot be found, it defaults backs to JUnit class order.
```
[the remark this answers is not in the document]
Comment (radcortez) [reply core/deployment/src/main/java/io/quarkus/deployment/dev/testing/TestConfig.java]: Done.

[the remark this answers is not in the document]
Comment (radcortez) [reply core/deployment/src/main/java/io/quarkus/deployment/dev/testing/TestConfig.java]: Yes, default is to not use an orderer. I've tried to improve the text to reflect that.

File: docs/src/main/asciidoc/getting-started-testing.adoc [hunk 4]
```diff
@@ -472,24 +472,9 @@ a bit slower, as it adds a shutdown/startup cycle to the test time, but gives a
...
-        <exclusion>
-            <groupId>io.quarkus</groupId>
-            <artifactId>quarkus-junit5-properties</artifactId>
-        </exclusion>
-    </exclusions>
-</dependency>
-----
+The behavior of this `ClassOrderer` is configurable via `application.properties` (see the source code or javadoc for
+more details). It can also be disabled entirely by setting another `ClassOrderer` that is provided by JUnit 5 or even
+your own custom one.
```
[the remark this answers is not in the document]
Comment (radcortez) [reply docs/src/main/asciidoc/getting-started-testing.adoc]: Added.

File: test-framework/junit5/src/main/java/io/quarkus/test/junit/util/QuarkusTestProfileAwareClassOrderer.java (hunk 2 above)
Comment (radcortez) [reply test-framework/junit5/src/main/java/io/quarkus/test/junit/util/QuarkusTestProfileAwareClassOrderer.java:96]: Removed.

**Comment by holly-cummins** [comment]
> > The `LauncherInterceptor` is invoked a few times during test launch, but the one I've been looking at is before discovery. The class orderer is invoked after discovery. I'm not sure if that would affect the success of what you're doing.
> 
> I had a better look, and I'm now using a `LauncherSessionListener` to implement the logic, which seems to execute after the first interceptor call. It works just as well. This should allow you to change any Classloader stuff before we set the configuration (where we need the Classloader we want to use).

Oh, even better! Thanks for thinking about that interaction.

**Comment by geoand** [comment]
Can we rebase this to see if the CI issue persists?

**Comment by radcortez** [comment]
Yes, done!

**Comment by geoand** [comment]
🙏

**Pull request merged**